// ng2-bootstrap/dist/buttons/button-radio.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
/// <reference path="../../tsd.d.ts" />
var angular2_1 = require('angular2/angular2');
var ButtonRadio = (function (_super) {
    __extends(ButtonRadio, _super);
    function ButtonRadio(cd, renderer, elementRef) {
        _super.call(this, cd, renderer, elementRef);
    }
    ButtonRadio.prototype.onInit = function () {
        this.uncheckable = typeof this.uncheckable !== 'undefined';
    };
    Object.defineProperty(ButtonRadio.prototype, "isActive", {
        get: function () {
            return this.btnRadio === this.value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ButtonRadio.prototype, "value", {
        get: function () {
            return this.cd.viewModel;
        },
        set: function (value) {
            this.cd.viewModel = value;
        },
        enumerable: true,
        configurable: true
    });
    ButtonRadio.prototype.writeValue = function (value) {
        this.value = value;
    };
    ButtonRadio.prototype.onClick = function () {
        if (this.uncheckable && this.btnRadio === this.value) {
            return this.cd.viewToModelUpdate(null);
        }
        this.cd.viewToModelUpdate(this.btnRadio);
    };
    ButtonRadio = __decorate([
        angular2_1.Directive({
            selector: '[btn-radio][ng-model]',
            properties: ['btnRadio', 'uncheckable'],
            host: {
                '(click)': 'onClick()',
                '[class.active]': 'isActive'
            }
        }),
        __param(0, angular2_1.Self()), 
        __metadata('design:paramtypes', [angular2_1.NgModel, angular2_1.Renderer, angular2_1.ElementRef])
    ], ButtonRadio);
    return ButtonRadio;
})(angular2_1.DefaultValueAccessor);
exports.ButtonRadio = ButtonRadio;
