// ng2-bootstrap/dist/datepicker/datepicker.js
/// <reference path="../../tsd.d.ts" />
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var angular2_1 = require('angular2/angular2');
var moment = require('moment');
var datepicker_inner_1 = require('./datepicker-inner');
var daypicker_1 = require('./daypicker');
var monthpicker_1 = require('./monthpicker');
var yearpicker_1 = require('./yearpicker');
var DatePicker = (function (_super) {
    __extends(DatePicker, _super);
    function DatePicker(cd, renderer, elementRef) {
        _super.call(this, cd, renderer, elementRef);
    }
    Object.defineProperty(DatePicker.prototype, "activeDate", {
        get: function () {
            return this._activeDate;
        },
        set: function (value) {
            this._activeDate = value;
            this.cd.viewToModelUpdate(moment(this.activeDate).toDate());
        },
        enumerable: true,
        configurable: true
    });
    DatePicker.prototype.onUpdate = function (event) {
        this.writeValue(event);
    };
    DatePicker.prototype.writeValue = function (value) {
        if (value) {
            if (typeof value !== 'Date') {
                value = new Date(value);
            }
            this.activeDate = value;
        }
    };
    DatePicker = __decorate([
        angular2_1.Component({
            selector: 'datepicker[ng-model], [datepicker][ng-model]',
            properties: [
                'datepickerMode',
                'minDate', 'maxDate',
                'dateDisabled', 'activeDate',
                'showWeeks', 'startingDay',
                'initDate',
                'minMode', 'maxMode',
                'formatDay', 'formatMonth', 'formatYear',
                'formatDayHeader', 'formatDayTitle', 'formatMonthTitle',
                'yearRange',
                'shortcutPropagation'
            ]
        }),
        angular2_1.View({
            template: "\n    <datepicker-inner [active-date]=\"activeDate\"\n                      (update)=\"onUpdate($event)\"\n                      [datepicker-mode]=\"datepickerMode\"\n                      [init-date]=\"initDate\"\n                      [min-date]=\"minDate\"\n                      [max-date]=\"maxDate\"\n                      [min-mode]=\"minMode\"\n                      [max-mode]=\"maxMode\"\n                      [show-weeks]=\"showWeeks\"\n                      [format-day]=\"formatDay\"\n                      [format-month]=\"formatMonth\"\n                      [format-year]=\"formatYear\"\n                      [format-day-header]=\"formatDayHeader\"\n                      [format-day-title]=\"formatDayTitle\"\n                      [format-month-title]=\"formatMonthTitle\"\n                      [starting-day]=\"startingDay\"\n                      [year-range]=\"yearRange\"\n                      [custom-class]=\"customClass\"\n                      [date-disabled]=\"dateDisabled\"\n                      [template-url]=\"templateUrl\"\n                      [shortcut-propagation]=\"shortcutPropagation\">\n      <daypicker tabindex=\"0\"></daypicker>\n      <monthpicker tabindex=\"0\"></monthpicker>\n      <yearpicker tabindex=\"0\"></yearpicker>\n    </datepicker-inner>\n    ",
            directives: [datepicker_inner_1.DatePickerInner, daypicker_1.DayPicker, monthpicker_1.MonthPicker, yearpicker_1.YearPicker, angular2_1.FORM_DIRECTIVES, angular2_1.CORE_DIRECTIVES]
        }),
        __param(0, angular2_1.Self()), 
        __metadata('design:paramtypes', [angular2_1.NgModel, angular2_1.Renderer, angular2_1.ElementRef])
    ], DatePicker);
    return DatePicker;
})(angular2_1.DefaultValueAccessor);
exports.DatePicker = DatePicker;
